HW2/P5/physics.pyx

```diff
@@ -69,18 +73,44 @@ cdef void sub_update(FLOAT[:, ::1] XY,
     ############################################################
     # SUBPROBLEM 2: use the grid values to reduce the number of other
     # objects to check for collisions.
-    for j in range(i + 1, count):
-        XY2 = &(XY[j, 0])
-        V2 = &(V[j, 0])
-        if overlapping(XY1, XY2, R):
-            # SUBPROBLEM 4: Add locking
-            if not moving_apart(XY1, V1, XY2, V2):
-                collide(XY1, V1, XY2, V2)
-
-            # give a slight impulse to help separate them
-            for dim in range(2):
-                V2[dim] += eps * (XY2[dim] - XY1[dim])
-
+    
+    #idea: change this for loop to iterate over all grid squares within 2 squares of XY1
+    # then check all j from those squares
+    grid_size = Grid.shape[0]
+    current_grid_x = <int>(XY1[0]/grid_spacing)
+    current_grid_y = <int>(XY1[1]/grid_spacing)
+    #make sure this object is in bounds
+    if ((current_grid_x >= 0) and (current_grid_x < grid_size) and (current_grid_y >= 0) and (current_grid_y < grid_size)): 
+        #get indexes of objects within 2 squares of XY1's grid square
+        for x in range(-2,3):
+            for y in range(-2,3):
+                idx = (x+2) * 5 + (y+2) # index in objects_within_2sqs
+                objects_within_2sqs[idx] = -1
+                grid_x = current_grid_x + x
+                grid_y = current_grid_y + y
+                #make sure this square is in the grid
+                if ((grid_x >= 0) and (grid_x < grid_size) and (grid_y >= 0) and (grid_y < grid_size)):
+                    #j is the index of the object in Grid[grid_x,grid_y]
+                    j = Grid[grid_x,grid_y]
+                    #make sure the object is not the same as object as i, and only store objects that have index greater than j, that way we don't account for collision twice
+                    if (j > i):
+                        objects_within_2sqs[idx] = j
+        for j_idx in range(25):# we set object_within_2sqs to have length 25 
```

Comment:
Seems a bit like you're using a "magic number", going through 25 iterations like this. Consider integrating this code into the nested for loop above